sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceManager.java

```diff
@@ -0,0 +1,282 @@
+import com.mesosphere.sdk.offer.TaskException;
+import com.mesosphere.sdk.scheduler.AbstractScheduler;
+import com.mesosphere.sdk.scheduler.DefaultScheduler;
+import com.mesosphere.sdk.scheduler.uninstall.UninstallScheduler;
+
+/**
+ * Handles adding/removing active services in a multi-service scheduler, and functions as the central source of truth
+ * for running services, including services which are in the process of uninstalling.
+ */
+public class MultiServiceManager {
+
+    private static final Logger LOGGER = LoggingUtils.getLogger(MultiServiceManager.class);
+
+    private final ReadWriteLock internalLock = new ReentrantReadWriteLock();
+    private final Lock rlock = internalLock.readLock();
+    private final Lock rwlock = internalLock.writeLock();
+
+    private final Map<String, AbstractScheduler> services = new HashMap<>();
+    private final Map<String, String> sanitizedServiceNames = new HashMap<>();
+
+    // Keeps track of whether we've had the registered callback yet.
+    // When a client is added, if we're already registered then invoke 'registered()' manually against that client
+    private boolean isRegistered;
+
+    public MultiServiceManager() {
+        this.isRegistered = false;
+    }
+
+    /**
+     * Returns the original names of currently available services.
+     */
+    public Collection<String> getServiceNames() {
+        Collection<String> serviceNames = new TreeSet<>(); // Alphabetical order
+        rlock.lock();
+        try {
+            serviceNames.addAll(services.keySet());
+        } finally {
+            rlock.unlock();
+        }
+        return serviceNames;
+    }
+
+    /**
+     * Returns the specified service by its original name, or an empty {@code Optional} if it's not found.
+     */
+    public Optional<AbstractScheduler> getService(String serviceName) {
+        rlock.lock();
+        try {
+            return Optional.ofNullable(services.get(serviceName));
+        } finally {
+            rlock.unlock();
+        }
+    }
+
+    /**
+     * Returns the specified service by its sanitized name, or an empty {@code Optional} if it's not found.
+     *
+     * A sanitized service name has had its slashes removed. For example,
+     * "{@code /path/to/service}" => "{@code path.to.service}". Sanitized names may be used in e.g. URLs that contain
+     * the service name.
+     */
+    public Optional<AbstractScheduler> getServiceSanitized(String sanitizedServiceName) {
+        rlock.lock();
+        try {
+            String originalName = sanitizedServiceNames.get(sanitizedServiceName);
+            if (originalName == null) {
+                // Unknown sanitized name
+                return Optional.empty();
+            }
+            return Optional.ofNullable(services.get(originalName));
+        } finally {
+            rlock.unlock();
+        }
+    }
+
+    /**
+     * Adds or replaces a running service according to the name in its ServiceSpec. Note: If the service was marked for
```

Comment:
What's the difference between add/replace?  Does replace == update?

Reply:
Add = start from scratch (new 'install')
Replace = start from previous state ('reconfigure')

Reply:
Changed this to 'Adds or updates'